app/src/main/java/io/homeassistant/companion/android/onboarding/discovery/HomeAssistantSearcher.kt

```diff
@@ -0,0 +1,86 @@
+package io.homeassistant.companion.android.onboarding.discovery
+
+import android.net.nsd.NsdManager
+import android.net.nsd.NsdServiceInfo
+import android.util.Log
+import java.net.URL
+import okio.internal.commonToUtf8String
+
+class HomeAssistantSearcher constructor(
+    private val nsdManager: NsdManager,
+    private val discoveryView: DiscoveryView
+) : NsdManager.DiscoveryListener {
+
+    companion object {
+        private const val SERVICE_TYPE = "_home-assistant._tcp"
+
+        private const val TAG = "HomeAssistantSearcher"
+    }
+
+    private var isSearching = false
+
+    fun beginSearch() {
+        if (isSearching)
+            return
+        isSearching = true
+        nsdManager.discoverServices(SERVICE_TYPE, NsdManager.PROTOCOL_DNS_SD, this)
+    }
+
+    fun stopSearch() {
+        if (!isSearching)
+            return
+        isSearching = false
+        nsdManager.stopServiceDiscovery(this)
+    }
+
+    // Called as soon as service discovery begins.
+    override fun onDiscoveryStarted(regType: String) {
+        Log.d(TAG, "Service discovery started")
+    }
+
+    override fun onServiceFound(foundService: NsdServiceInfo) {
+        // A service was found! Do something with it.
+        Log.d(TAG, "Service discovery success: $foundService")
+
+        if (foundService.serviceType.startsWith(SERVICE_TYPE)) {
```

Comment:
Consider using a guard clause for readability.

```kotlin
if (!foundService.serviceType.startsWith(SERVICE_TYPE)) { return; }
```

However, do we even need this check? We pass `SERVICE_TYPE` to `nsdManager.discoverServices`